src/NUnitFramework/framework/Attributes/RetryAttribute.cs

```diff
@@ -91,10 +91,10 @@ public override TestResult Execute(TestExecutionContext context)
 
                 while (count-- > 0)
                 {
-                    context.CurrentResult = innerCommand.Execute(context);
+                    innerCommand.ExecuteAndSetCurrentResult(context);
```

Comment:
The separate method to do two things seems wrong to me and makes this harder to read IME. What's wrong in principle with the idea that the __caller__ should decide what to do with the returned result? It's a pretty fundamental assumption in the `Command` structure and I worry about how invalidating that assumption will affect other things.

Reply:
To answer multiple comments at once: 

Everything I'm reading shows that the Command structure relies heavily on the most current result always being set to the execution context. I was confronted with this during my three "fix" commits in this PR. Every command which need to be aware of result status is responsible to **both** use try...catch (in case the inner command doesn't catch its own exceptions) **and also** check the inner command's result if it returns without exception to see if it had caught and recorded an exception. So whatever we do, we need a helper method to make sure these two tasks stay in sync.

Since many commands are tied so intimately to the execution context's CurrentResult property, I figured this was the shortest path to a safe place. My preference would definitely be to remove all side-effects from the Command structure and make the returned TestResult be the single source of truth but even then, delegating commands need a helper method to catch exceptions and merge them into the inner command's result. Does that make sense?

Reply:
It certainly makes for simpler code if you do it that way. However...

Nobody should be handling exceptions if they don't know what those exceptions mean. In the case of a user-provided BeforeAfterCommand, we don't actually know what an exception means if it once escapes from the Command. This tells me that each command should be responsible for handling its own exceptions. At a minimum, it should translate the exception into a proper result and return it. From your comments, I think we agree up to that point at least.

That, BTW, was how the command structure worked before some changes messed it up. I dislike having to have a try / catch in every command, but that's what I would have done initially - in keeping with my mantra of creating duplication before removing it. 😄 I have trouble with an approach based on first __imagining__ the duplication and then mentally removing it because I've watched so many teams mess up by doing exactly that. That's off-topic here, however.

Where I think we disagree is with the notion that the returned `TestResult` should also be set into the `TestExecutionContext`. The Command structure doesn't - or shouldn't - depend on this result being set correctly. Rather, the code in `WorkItem` depends on it. It needs to be set after an entire chain of commands executes. We can achieve that by having each command __exclusively__ make use of the result that is returned to it and not the current result. Doing that has the side effect that you can test Commands without requiring more than a dummy context.

That's my view for your consideration.

Reply:
Well, I'm not talking about imagined duplication. [SimpleWorkItem](https://github.com/nunit/nunit/blob/3c028034e717c79bf00661c5852b358f9097d2e1/src/NUnitFramework/framework/Internal/Execution/SimpleWorkItem.cs#L57-L75) [BeforeAndAfterTestCommand](https://github.com/nunit/nunit/blob/3c028034e717c79bf00661c5852b358f9097d2e1/src/NUnitFramework/framework/Internal/Commands/BeforeAndAfterTestCommand.cs#L55-L68)
And notice what we don't have in BeforeTestCommand and AfterTestCommand...

Great! We don't actually disagree there! The problem is that we currently do everything in terms of the execution context's CurrentResult, so I say we should either double down on what we're already doing and make it safe and consistent (which is what you see here) or we should rewrite much of the Command structure. I considered my choice of sticking with the current pattern to be a reasonable compromise but I'm not opposed to an overhaul. I just don't think we should leave it inconsistent. To me, that's asking for misunderstandings as future features and fixes are implemented.

Reply:
Please don't misunderstand! I'm not disagreeing with your vision of making the command structure purer. That would be my ideal for sure! I'm used to being told that I need to be more pragmatic and not to make such big changes. 😜 But it's urgent to me to move to *some* safe place where there is a pit of success for us and even for NUnit consumers when implementing test commands. Perhaps not this PR then.

Reply:
Right... I could hypothetically review this PR in two modes: (1) a small fix that got too big (2) a restructure that didn't go far enough. Maybe we'll opt for (3) a fix that initiates a future restructure.

While I abhor the phrase "pit of success" I agree that we should make it relatively easy for a user to write a Command, that being one of our supported extensions. It seems easier to me if the user merely has to return a result and not handle errors. In fact, even if the user is supposed to handle errors, we should probably allow for user error and handle them ourselves.